packages/icon-button/src/icon-button-base.ts

```diff
@@ -0,0 +1,101 @@
+/**
+@license
+Copyright 2018 Google Inc. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import {BaseElement, html, property, Foundation, Adapter, query, observer} from '@material/mwc-base/base-element.js';
+import MDCIconButtonToggleFoundation from '@material/icon-button/foundation.js';
+import {ripple} from '@material/mwc-ripple/ripple-directive.js';
+
+export interface IconButtonFoundation extends Foundation {
+  handleClick(): void;
+  toggle(isOn?: boolean): void;
+}
+
+export declare var IconButtonFoundation: {
+  prototype: IconButtonFoundation;
+  new (adapter: Adapter): IconButtonFoundation;
+}
+
+export abstract class IconButtonBase extends BaseElement {
+
+  protected mdcFoundationClass: typeof IconButtonFoundation = MDCIconButtonToggleFoundation;
+
+  protected mdcFoundation!: IconButtonFoundation;
+
+  @query('.mdc-icon-button')
+  protected mdcRoot!: HTMLElement;
+
+  @property({type: String})
+  label = '';
+
+  @property({type: Boolean, reflect: true})
+  disabled = false;
+
+  @property({type: String})
+  icon = '';
+
+  @property({type: String})
+  offIcon = '';
```

Comment:
We discussed making the `offIcon` not optional, but if it is not, how is the behavior different? Would we throw if there's no `offIcon`? That would be strange.

Reply:
If there is no `offIcon`, the state never toggles and we do not emit an event.